Extract table components from My Yoast

IreneStr opened the pull request:
### Summary
Parts from @jcomack's PR:
- Extracts the Cell (renamed from 'Column', see below), ListTable and Row components (and their variants) from My Yoast.
- Separates ZebrafiedTable component from the ListTable component.
- Dependency additions: lodash, debug, prop-types, styled-components.

Changes on top of @jcomack's PR:
- Adds tests
- Renames 'Column' to 'Cell'
- Makes sure zebrafying is working when there is only one child
- Adds `padding-left: 0` to spans in RowMobileCollapse, to prevent padding differences between the first cell and the other cells.

### Changelog
Changelog Entry: Introduce Cell, Row and ListTable components.

### Testing
Paste the following code into SnippetPreview.js, and play around with the props etc.
````js
import React from "react";
import styled from "styled-components";

import { CellPrimary, CellIcon, CellMinWidth, CellFixedWidth, responsiveHeaders } from "../../basic/Table/Cell";
import { Row, RowResponsiveWrap } from "../../basic/Table/Row";
import { ListTable, ZebrafiedListTable, makeFullWidth } from "../../basic/Table/ListTable";

export const SnippetPreviewDiv = styled.div`
	height: 700px;
	width: 100%;
	background-color: white;
`;

let MinWidthCellResponsive = makeFullWidth( responsiveHeaders( CellMinWidth ) );
let rowProps = [];

/**
 * Returns the SnippetPreview component.
 *
 * @returns {ReactElement} The SnippetPreview component.
 */
export default function SnippetPreview() {
	return(
		<SnippetPreviewDiv>
			<br /><br /><br />
			<ListTable>
				<RowResponsiveWrap { ...rowProps }>
					<CellPrimary headerLabel="Header 1">
						This is a primary cell
					</CellPrimary>
					<CellMinWidth headerLabel="Header 2">
						This is a min width cell
					</CellMinWidth>
					<CellFixedWidth headerLabel="Header 3">
						This is a fixed width cell
					</CellFixedWidth>
					<CellMinWidth headerLabel="Header 4">
						This is a min width cell
					</CellMinWidth>
					<CellIcon headerLabel="X">
						Icon
					</CellIcon>
					<MinWidthCellResponsive headerLabel="Header 5">
						This is a responsive min width cell with header
					</MinWidthCellResponsive>
				</RowResponsiveWrap>
			</ListTable>
			<br /><br /><br /><br /><br /><br />
			<ZebrafiedListTable>
				<Row { ...rowProps }>
					<CellPrimary ellipsis={ true } headerLabel="Header 1">
						This is a primary cell
					</CellPrimary>
					<CellMinWidth ellipsis={ true } headerLabel="Header 2">
						This is a min width cell
					</CellMinWidth>
					<CellFixedWidth ellipsis={ true } headerLabel="Header 3">
						This is a fixed width cell
					</CellFixedWidth>
					<CellMinWidth ellipsis={ true } headerLabel="Header 4">
						This is a min width cell
					</CellMinWidth>
					<CellIcon headerLabel="X">
						Icon
					</CellIcon>
				</Row>
				<Row { ...rowProps }>
					<CellPrimary ellipsis={ true } headerLabel="Header 1">
						This is a primary cell
					</CellPrimary>
					<CellMinWidth ellipsis={ true } headerLabel="Header 2">
						This is a min width cell
					</CellMinWidth>
					<CellFixedWidth ellipsis={ true } headerLabel="Header 3">
						This is a fixxed width cell
					</CellFixedWidth>
					<CellMinWidth ellipsis={ true } headerLabel="Header 4">
						This is a min width cell
					</CellMinWidth>
					<CellIcon headerLabel="X">
						Icon
					</CellIcon>
				</Row>
				<Row { ...rowProps }>
					<CellPrimary ellipsis={ true } headerLabel="Header 1">
						This is a primary cell
					</CellPrimary>
					<CellMinWidth ellipsis={ true } headerLabel="Header 2">
						This is a min width cell
					</CellMinWidth>
					<CellFixedWidth ellipsis={ true } headerLabel="Header 3">
						This is a fixxed width cell
					</CellFixedWidth>
					<CellMinWidth ellipsis={ true } headerLabel="Header 4">
						This is a min width cell
					</CellMinWidth>
					<CellIcon headerLabel="X">
						Icon
					</CellIcon>
				</Row>
			</ZebrafiedListTable>
		</SnippetPreviewDiv>
	);
}
````

Fixes #210

afercia commented:
CR done 👍 

Moving back because @IreneStr informed me she's going to add some more tests.

Some considerations, more related to acceptance than CR:

On My Yoast, everything inherits `box-sizing: border-box` while here it's the default `content-box`; this actually affects the size of the flex items. I'd consider to switch everything to `border-box` but for performance reasons  this should be done at the root level using the universal selector `*` and not in the components.

I'd suggest to rename the cells. While the Rows are named based on {Row}+{type}:
```
Row
RowMobileCollapse
```

The cells are named differently and maybe it would be better to prepend "what" they are followed by the type. Currently:
```
BaseCell
PrimaryCell
FixedWidthCell
MinWidthCell
IconCell
```

Maybe better: CellBase, CellPrimary, etc.

Media queries breakpoints are hardcoded. On My Yoast instead, we're using some values from a config file. I'd consider to do the same thing here, maybe in a separate PR.

I'd consider to rename `RowMobileCollapse` as it doesn't really tell me what it does. Actually, it allows its flex-items children to wrap in a new line (`flex-wrap: wrap;`) in the responsive view. So maybe `RowResponsiveWrap` or something similar would be more appropriate.

On My Yoast, there are a few cases where we make variants in other files, for example: `RowMobileCollapseNoMinHeight` and `ColumnFixedWidthResponsive`. Not sure if they should be included in this PR or be considered exceptions.

Notes on usage:
- the Primary Cell is meant to be the main cell in a row, and there should be just one in a row
- the cell with responsive header is mostly meant to be used together with `makeFullWidth` to make it full width in the responsive view; thus it makes sense to use it just within a responsive row (the one that can wrap)
- "headers" should always be short and never go in two lines because they currently use `line-height: 0;`, maybe we should review this (see screenshot below)
- icon cells are meant to contain only icons, so they shouldn't use ellipsis
- maybe these notes on usage should be included in a readme?

![screen shot 2017-08-10 at 14 59 58](https://user-images.githubusercontent.com/1682452/29171505-12315c44-7ddd-11e7-8b9b-b28ea29f6b97.png)

The highlighted cell header text is "Responsive Cell Header". When it's too long and there's not enough space, it technically goes in two lines but visually overlays itself because of `line-height: 0;`


Based on the above, I'd suggest to test with the following new content:
```
import React from "react";
import styled from "styled-components";

import { PrimaryCell, IconCell, MinWidthCell, FixedWidthCell, responsiveHeaders } from "../../basic/Table/Cell";
import { Row, RowMobileCollapse } from "../../basic/Table/Row";
import { ListTable, ZebrafiedListTable, makeFullWidth } from "../../basic/Table/ListTable";

export const SnippetPreviewDiv = styled.div`
	height: 700px;
	width: 100%;
	background-color: white;
`;

let MinWidthCellResponsive = makeFullWidth( responsiveHeaders( MinWidthCell ) );
let rowProps = [];

/**
 * Returns the SnippetPreview component.
 *
 * @returns {ReactElement} The SnippetPreview component.
 */
export default function SnippetPreview() {
	return(
		<SnippetPreviewDiv>
			<br /><br /><br />
			<ListTable>
				<RowMobileCollapse { ...rowProps }>
					<PrimaryCell headerLabel="Header 1">
						This is a primary cell
					</PrimaryCell>
					<MinWidthCell headerLabel="Header 2">
						This is a min width cell
					</MinWidthCell>
					<FixedWidthCell headerLabel="Header 3">
						This is a fixed width cell
					</FixedWidthCell>
					<MinWidthCell headerLabel="Header 4">
						This is a min width cell
					</MinWidthCell>
					<IconCell headerLabel="X">
						Icon
					</IconCell>
					<MinWidthCellResponsive headerLabel="Header 5">
						This is a responsive min width cell with header
					</MinWidthCellResponsive>
				</RowMobileCollapse>
			</ListTable>
			<br /><br /><br /><br /><br /><br />
			<ZebrafiedListTable>
				<Row { ...rowProps }>
					<PrimaryCell ellipsis={ true } headerLabel="Header 1">
						This is a primary cell
					</PrimaryCell>
					<MinWidthCell ellipsis={ true } headerLabel="Header 2">
						This is a min width cell
					</MinWidthCell>
					<FixedWidthCell ellipsis={ true } headerLabel="Header 3">
						This is a fixed width cell
					</FixedWidthCell>
					<MinWidthCell ellipsis={ true } headerLabel="Header 4">
						This is a min width cell
					</MinWidthCell>
					<IconCell headerLabel="X">
						Icon
					</IconCell>
				</Row>
				<Row { ...rowProps }>
					<PrimaryCell ellipsis={ true } headerLabel="Header 1">
						This is a primary cell
					</PrimaryCell>
					<MinWidthCell ellipsis={ true } headerLabel="Header 2">
						This is a min width cell
					</MinWidthCell>
					<FixedWidthCell ellipsis={ true } headerLabel="Header 3">
						This is a fixxed width cell
					</FixedWidthCell>
					<MinWidthCell ellipsis={ true } headerLabel="Header 4">
						This is a min width cell
					</MinWidthCell>
					<IconCell headerLabel="X">
						Icon
					</IconCell>
				</Row>
				<Row { ...rowProps }>
					<PrimaryCell ellipsis={ true } headerLabel="Header 1">
						This is a primary cell
					</PrimaryCell>
					<MinWidthCell ellipsis={ true } headerLabel="Header 2">
						This is a min width cell
					</MinWidthCell>
					<FixedWidthCell ellipsis={ true } headerLabel="Header 3">
						This is a fixxed width cell
					</FixedWidthCell>
					<MinWidthCell ellipsis={ true } headerLabel="Header 4">
						This is a min width cell
					</MinWidthCell>
					<IconCell headerLabel="X">
						Icon
					</IconCell>
				</Row>
			</ZebrafiedListTable>
		</SnippetPreviewDiv>
	);
}
```

IreneStr commented:
### In this PR
- Renamed the Cell components
- Added a Table Read Me
- Renamed 'RowMobileCollapse' to 'RowResponsiveWrap'
- Updated tests

I've also updated the test instructions based on the new Cell names.

### Separate issue
I've created a separate issue for switching to `border-box` and creating a config file with media query breakpoints: #220

### Other

- Components like RowMobileCollapseNoMinHeight and ColumnFixedWidthResponsive are either just a component with `makeFullWidth` applied to it, or no longer relevant due to the slightly more generic architecture of this table components in comparison to My Yoast.

boblinthorst commented:
CR: seemed fine, with the seperate issues created for the problems pointed out by Andrea taken into account. Acceptance: seemed to have some issues aligning the cells which I thought didn't look all that great:

<img width="1662" alt="table_component_lining" src="https://user-images.githubusercontent.com/11849359/29215666-2bef06de-7eac-11e7-98e3-941d8497481d.png">

After discussing this with @IreneStr , A seperate issue has been created to address this. So with that out of the way. Acceptance seemed ok 👍 .
